File: 473/473.cpp
/*
Runtime: 1444 ms, faster than 36.59% of C++ online submissions 
*/

class Solution {
public:
    bool flag;
    
    // calculate the sum of the vector
    int computeSum(vector<int> v) {
        int ret = 0, length = v.size();
        for(int i=0;i<length;i++) {
            ret += v[i];
        }
        return ret;
    }
    
    void dfs(vector<int> sbsum, int width, vector<int> lst, int index) {
        if(index == lst.size()) {
            flag = true;
            return;
        }
        if(flag)
            return;
        
        for(int i=0;i<4;i++) {
            if(sbsum[i]+lst[index] <= width) {
                sbsum[i] += lst[index];
                dfs(sbsum, width, lst, index+1);
                sbsum[i] -= lst[index];
            }
        }
    }
    
    bool makesquare(vector<int>& nums) {
        flag = false;
        int length = nums.size();
        int totsum = computeSum(nums);
        
        if(totsum<=0 || totsum%4!=0)
            return false;
        int width = totsum / 4;
        vector<int> sbsum = {0, 0, 0, 0};
        // sort the list. 
        // rbegin(): the first element from the end.
        sort(nums.rbegin(), nums.rend());
        dfs(sbsum, width, nums, 0);
        return flag;
    }
};

File: 473/473.py
'''
We should divide the list into 4 subsets. Each subset has the same sum.
So for each element in the list, consider whether it belongs to the first
subset? the second subset?... etc. Depth-first Search.

Tip: It will be quicker if I sort the list.
'''

class Solution:
    flag = False
    
    def dfs(self, sbsum, width, lst, index):
        if index == len(lst):
            self.flag = True
            return
        if self.flag:
            return
            
        for i in range(4):
            if sbsum[i] + lst[index] <= width:
                sbsum[i] += lst[index]
                self.dfs(sbsum, width, lst, index+1)
                sbsum[i] -= lst[index]
    
    def makesquare(self, nums: 'List[int]') -> 'bool':
        totlength = sum(nums)
        if totlength%4!=0 or totlength<=0:
            return False
        # get the width of square
        width = totlength // 4
        if max(nums) > width:
            return False
        
        # It will get a TLE if I don't sort it
        nums.sort(reverse=True)
                
        self.flag = False
        sbsum = [0, 0, 0, 0]        # sum of each subset
        self.dfs(sbsum, width, nums, 0)
        
        return self.flag